doc/decisions/1_in_discussion/copy_on_write.md

```diff
@@ -327,12 +324,101 @@ API notes:
 - `ksLookup()`:
   - if the `DEL` or `POP` flag is specified, do the COW-stuff as described multiple times now in the operations above
 
-#### Optimizations
+#### Reference Counting
+
+We need reference counting for the internal COW datastructures.
+For the beginning, we can just do it the same way reference counting currently works for `Key` and `KeySet`.
+One tweak though is that the refcount should never be 0, as this does not make sense for internal datastructures.
+
+This means we always increment the refcount after creation and always decrement before deletion, so that the refcount is never zero.
+An example implementation is shown below:
+
+```c
+static void keySetValue(Key * key, void * value, size_t size) {
+  // [...] removal of current value from key
+  struct _KeyData data = keyDataNew (value, size);
+  keyDataIncRef (data);
+  key->data = data;
+}
+
+static void keyDel(Key * key) {
+  keyDataDecRef (key->data);
+  keyDataDel (key->data);
+  // [...] other cleanup
+}
+```
+
+#### Variation 1 - RcBuffer
+
+Instead of using different structs for `_KeyData`, and `_KeyName` use a more generic struct for reference counting.
+This would avoid some duplication on the reference counting code for the key.
+Keysets will still have their own data struct, as it contains more than just a pointer and a size.
+
+```c
+typedef struct {
+	void * data;
+	size_t size;
+	uint16_t refs;
+} RcBuffer;
+
+struct _Key {
+	RcBuffer * uname;
+	RcBuffer * ename; // will be removed soon
+	RcBuffer * value;
+	
+	KeySet * meta;
+	keyflag_t flags;
+	uint16_t refs;
+};
+```
+
+#### Possible Edge Cases
+
+In general, it should be possible to always do copy-on-write.
+From a users perspective, copy-on-write copies of a key (and a keyset) should behave the same.
+There is, however, one edge case: the user modifying the value of a key directly.
+This is shown in the following example:
+
+```c
+Key * key;
+struct foo myFoo = {
+  .x = 0
+};
+keySetBinary (key, &myFoo, sizeof(myFoo));
+
+Key * dup = keyDup (key);
 
-This approach requires more allocations than previously.
-We have not fully benchmarked whether this is a big issue.
-One optimization could be an expanding "pool" of `_KeySetData`, `_KeyData` and `_KeyName`.
-We could then allocate multiple of them at the same time, and borrow and give back instance from and to the pool.
+((struct foo *)keyValue (key))->x = 1;
+
+// with COW
+assert (((struct foo *)keyValue (dup))->x == 1);
+// without COW
+assert (((struct foo *)keyValue (dup))->x == 0);
+```
+
+This edge case can be accounted for by providing a function, i.e. `keyDetach`, that forces that the key has its very own copy of the data.
```

Comment:
I don't think we need this.

Reply:
The function is not just for this case, but for the general case, when you want a non-COW copy for whatever reason. However, I did already say that I'd probably put the function into `libelektra-extras` or similar, if we add it at all, since it would just be something like:

```c
// WARNING: pseudo-code, actual code slightly more complicated
static inline void keyDetach (Key * key) {
    keySetName (key, keyName (key));
    keySetValue (key, keyValue (key));
    // maybe do something for metadata, but I don't think it's needed
}
```

Reply:
No, we need this. During the implementation I had to create such functions for both Key and KeySet to avoid code duplication. We can argue whether or not they should be part of the public API though.

Reply:
To come to a conclusion here:

We can leave the note in the file. It just says this would be a possible solution for the edge case. It doesn't say that we will introduce this.

Whether we do need it can be discussed in the PR implementing everything, where we see the code. If we really do need it, we can then discuss whether we make it public API and via which library.

Reply:
```suggestion
This edge case can be accounted for by providing a private function `keyDetach()`, that forces that the key has its very own copy of the data.
```

Yes, let us introduce it as private API and we can later decide if it must be public if we know more about the use cases of Elektra's API.